src/rust/sqs-executor/src/s3_event_emitter.rs

```diff
@@ -0,0 +1,321 @@
+
+        for event in events {
+            let on_emit = self.on_emit.clone();
+            let output_bucket = self.output_bucket.clone();
+            let key = (self.key_fn)(&event);
+            let s3 = self.s3.clone();
+            let mut metric_reporter = self.metric_reporter.clone();
+            let put_object = async move {
+                tracing::debug!("uploading event to: {} {}", output_bucket, key);
+                println!("uploading event to: {} {}", output_bucket, key);
+                let (res, ms) = s3
+                    .put_object(PutObjectRequest {
+                        body: Some(event.into()),
+                        bucket: output_bucket.clone(),
+                        key: key.clone(),
+                        ..Default::default()
+                    })
+                    .timed()
+                    .await;
+                metric_reporter
+                    .histogram(
+                        "s3_event_emitter.emit_event.ms",
+                        ms as f64,
+                        &[tag("success", res.is_ok())],
+                    )
+                    .unwrap_or_else(|e| error!("s3_event_emitter.emit_event.ms: {:?}", e));
+
+                match res {
+                    Ok(res) => {
+                        tracing::debug!("PutObject succeeded: {:?}", res);
+                        println!("PutObject succeeded: {:?}", res);
+                        on_emit
+                            .event_notification(output_bucket, key)
+                            .await
+                            .map_err(|e| S3EventEmitterError::OnEmitError(e))
+                            .tap_err(|e| error!("event_notification failed: {:?}", e))?;
+                        Ok(())
+                    }
+                    Err(e) => {
+                        tracing::warn!("PutObject failed: {:?}", e);
+                        println!("PutObject failed: {:?}", e);
+                        Err(S3EventEmitterError::from(e))
+                    }
+                }
+            };
+            event_uploads.push(put_object);
+        }
+        let event_uploads = tokio::time::timeout(
+            Duration::from_secs(3),
+            futures::future::join_all(event_uploads),
+        )
+        .await?;
+
+        for e in event_uploads.into_iter() {
+            e?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum S3NotificationError {
+    #[error("PublishError")]
+    PublishError(#[from] RusotoError<SendMessageError>),
+}
+
+impl CheckedError for S3NotificationError {
+    fn error_type(&self) -> Recoverable {
+        Recoverable::Transient
+    }
+}
+
+#[async_trait]
+pub trait OnEventEmit {
+    type Error: CheckedError;
+    async fn event_notification(&self, bucket: String, key: String) -> Result<(), Self::Error>;
+}
+
+pub struct S3ToSqsEventNotifier<S>
+where
+    S: Sqs + Clone + Send + Sync + 'static,
```

Comment:
not a blocker, more of a "max learning rust" question: is there a way to alias this S so that it doesn't need to be repeated everywhere? like

type SBound = Sqs + Clone + Send + Sync + 'static;

Reply:
You would use a marker trait.

```
trait SBound: Sqs + Clone + Send + Sync + 'static {}
```